contracts/src/StakeTable.sol

```diff
@@ -374,60 +397,134 @@ contract StakeTable is AbstractStakeTable {
         uint256 balance = node.balance;
         if (balance == 0) {
             revert InsufficientStakeBalance(0);
         }
 
-        // Verify that the validator can sign for that blsVK
-        bytes memory message = abi.encode(msg.sender);
-        BLSSig.verifyBlsSig(message, blsSig, blsVK);
-
         // Verify that the exit escrow period is over.
         if (currentEpoch() < node.exitEpoch + exitEscrowPeriod(node)) {
             revert PrematureWithdrawal();
         }
 
         // Delete the node from the stake table.
-        delete nodes[key];
+        delete nodes[msg.sender];
 
         // Transfer the balance to the node's account.
         SafeTransferLib.safeTransfer(ERC20(tokenAddress), node.account, balance);
 
         return balance;
     }
 
+    /// @notice Update the consensus keys for a validator
+    /// @dev This function is used to update the consensus keys for a validator
+    /// @dev This function can only be called by the validator itself when it's not in the exit
+    /// queue
+    /// @dev The validator will need to give up either its old BLS key and/or old Schnorr key
+    /// @dev The validator will need to provide a BLS signature over the new BLS key
+    /// @param newBlsVK The new BLS verification key
+    /// @param newSchnorrVK The new Schnorr verification key
+    /// @param newBlsSig The BLS signature that the account owns the new BLS key
+    /// TODO: consider emitting the changed keys
+    function updateConsensusKeys(
+        BN254.G2Point memory newBlsVK,
+        EdOnBN254.EdOnBN254Point memory newSchnorrVK,
+        BN254.G1Point memory newBlsSig
+    ) external override {
+        Node memory node = nodes[msg.sender];
+
+        // Verify that the node is already registered.
+        if (node.account == address(0)) {
+            revert NodeNotRegistered();
+        }
+
+        // Verify that the node is not in the exit queue
+        if (node.exitEpoch != 0) {
+            revert ExitRequestInProgress();
+        }
+
+        // The staker does not provide a key change
+        if (
+            (
+                _isEqualBlsKey(newBlsVK, node.blsVK)
+                    && EdOnBN254.isEqual(newSchnorrVK, node.schnorrVK)
+            )
+                || (
+                    _isEqualBlsKey(
+                        newBlsVK,
+                        BN254.G2Point(
+                            BN254.BaseField.wrap(0),
+                            BN254.BaseField.wrap(0),
+                            BN254.BaseField.wrap(0),
+                            BN254.BaseField.wrap(0)
+                        )
+                    ) && EdOnBN254.isEqual(newSchnorrVK, EdOnBN254.EdOnBN254Point(0, 0))
+                )
+        ) {
+            revert NoKeyChange();
+        }
+
+        // Update the node's schnorr key once it's not the same as the old one and it's nonzero
+        if (
+            !EdOnBN254.isEqual(newSchnorrVK, node.schnorrVK)
```

Comment:
For the `EdOnBn254` and `BN254` types, can we use `using ... for ...` to make the code a bit more readable?